hbase-server/src/main/java/org/apache/hadoop/hbase/procedure2/store/region/HFileProcedurePrettyPrinter.java

```diff
@@ -0,0 +1,174 @@
+import org.apache.hbase.thirdparty.org.apache.commons.cli.Option;
+import org.apache.hbase.thirdparty.org.apache.commons.cli.OptionGroup;
+
+import org.apache.hadoop.hbase.shaded.protobuf.generated.ProcedureProtos;
+
+/**
+ * A tool to dump the procedures in the HFiles.
+ * <p/>
+ * The different between this and {@link org.apache.hadoop.hbase.io.hfile.HFilePrettyPrinter} is
+ * that, this class will decode the procedure in the cell for better debugging. You are free to use
+ * {@link org.apache.hadoop.hbase.io.hfile.HFilePrettyPrinter} to dump the same file as well.
+ */
+@InterfaceAudience.LimitedPrivate(HBaseInterfaceAudience.TOOLS)
+@InterfaceStability.Evolving
+public class HFileProcedurePrettyPrinter extends AbstractHBaseTool {
+
+  private Long procId;
+
+  private List<Path> files = new ArrayList<>();
+
+  private final PrintStream out;
+
+  public HFileProcedurePrettyPrinter() {
+    this(System.out);
+  }
+
+  public HFileProcedurePrettyPrinter(PrintStream out) {
+    this.out = out;
+  }
+
+  @Override
+  protected void addOptions() {
+    addOptWithArg("w", "seekToPid", "Seek to this procedure id and print this procedure only");
+    OptionGroup files = new OptionGroup();
+    files.addOption(new Option("f", "file", true,
+      "File to scan. Pass full-path; e.g. hdfs://a:9000/MasterProcs/master/procedure/p/xxx"));
+    files.addOption(new Option("a", "all", false, "Scan the whole procedure region."));
+    options.addOptionGroup(files);
+  }
+
+  private void addAllHFiles() throws IOException {
+    Path masterProcDir =
+      new Path(CommonFSUtils.getWALRootDir(conf), RegionProcedureStore.MASTER_PROCEDURE_DIR);
+    Path tableDir = CommonFSUtils.getTableDir(masterProcDir, RegionProcedureStore.TABLE_NAME);
+    FileSystem fs = tableDir.getFileSystem(conf);
+    Path regionDir =
+      fs.listStatus(tableDir, p -> RegionInfo.isEncodedRegionName(Bytes.toBytes(p.getName())))[0]
+        .getPath();
+    List<Path> regionFiles = HFile.getStoreFiles(fs, regionDir);
+    files.addAll(regionFiles);
+  }
+
+  @Override
+  protected void processOptions(CommandLine cmd) {
+    if (cmd.hasOption("w")) {
+      String key = cmd.getOptionValue("w");
+      if (key != null && key.length() != 0) {
+        procId = Long.parseLong(key);
+      } else {
+        throw new IllegalArgumentException("Invalid row is specified.");
+      }
+    }
+    if (cmd.hasOption("f")) {
+      files.add(new Path(cmd.getOptionValue("f")));
+    }
+    if (cmd.hasOption("a")) {
+      try {
+        addAllHFiles();
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+    }
+  }
+
+  private void printCell(Cell cell) throws IOException {
+    out.print("K: " + CellUtil.getCellKeyAsString(cell,
+      c -> Long.toString(Bytes.toLong(c.getRowArray(), c.getRowOffset(), c.getRowLength()))));
+    if (cell.getType() == Cell.Type.Put) {
+      if (cell.getValueLength() == 0) {
+        out.println(" V: mark deleted");
```

Comment:
What does 'mark deleted' mean?

Reply:
You can see the comments of RegionProcedureStore. For preserving the max procedure id, we can not delete a procedure directly. Instead, we just put an empty array to overwrite the procedure value, to indicate that the procedure has been deleted, This is a 'mark deleted'.